text/0000-deref-move.md

```diff
@@ -0,0 +1,322 @@
+This can occur either because an explicit dereference is used in a context where
+a value is to be moved, or because autoderef selects a method taking `self` (or
+`Box<self>`) as a parameter. In both cases, when evaluating whether to call
+`deref_move`, the compiler will look at whether `Target` implements `Copy`. If
+it does not, then `deref_move` is called to move the target out of the smart
+pointer. If it does, then instead the compiler will only call `deref` and then
+copy out of the returned reference. For generic types, the compiler only checks
+if a `Copy` bound is available at the use site; it is not done during
+monomorphization.
+
+For explicit dereference, the changes are straightforward. Any explicit
+dereference of a `DerefMove` type becomes a movable expression, just as
+dereference of `Box` is today.
+
+For autodereference: 
+
+* In a function call expression, the callee operand may be dereferenced and
+  moved if the `Target` implements `FnOnce` and no reference-based call trait.
+* In a method call expression, the receiver operand may be dereferenced and
+  moved if the function selected by overload resolution takes `self` or
+  `Box<self>` as a parameter (in the latter case, this would require a smart
+  pointer containing a `Box<Self>`, of course).
+* In a field access expression, the left-hand operand may be dereferenced and
+  moved if the result of the expression is moved. In this case, the move is
+  always entire; partial move is not supported when a dereference is necessary.
+* An index expression can never move its left operand, so it is unaffected.
+* A borrow expression can never move its operand, so it is unaffected.
+
+No change is made to overload resolution of method calls. This means that calls
+which are invalid today due to preferring `T` over `&T` may now be legal,
+selecting to move out of the smart pointer rather than giving an error.
+
+There is an ambiguity in a method call expression when the `Target` type of the
+receiver is a reference, and that reference type has a method taking `self`. In
+this case, autodereference could either dereference via `deref_move` or via
+`deref` or `defer_mut`. In this case, it will prefer the latter (for `deref`, it
+is a special case of the `Copy` rule above, but for `deref_mut` it is not) to
+avoid unnecessarily moving out of the receiver.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Nobody understands autoderef already and this makes it worse
+
+Or at least, it would if `Box` wasn't already complicating it in the same way.
+
+## Copy behaviour is unintuitive in generic cases
+
+If a type has nontrivial/side-effectful `defer_move`, it may be slightly
+confusing why it is not used for `Copy` types in non-generic code, but is used
+in generic code.
+
+## Breaking change to an edge case of `Box<&mut T>`
+
+Currently, a `Box<&mut T>` variable [can be
+dereferenced](https://play.rust-lang.org/?gist=5b3f046ecf29da8f00fde3bd1d2a3df4&version=nightly&mode=debug)
```

Comment:
I would prefer if you inlined this example, so that these documents are less prone to bitrot.